compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp

```diff
@@ -1803,6 +1806,221 @@ std::optional<StringRef> LoweringConfigAttr::getLoweringStrategy() const {
   }
   return std::nullopt;
 }
+constexpr StringLiteral kWorkgroupReOrderingStrategyName =
+    "workgroup_reordering_strategy";
+
+::mlir::iree_compiler::IREE::Codegen::WorkgroupReorderingAttrInterface
+LoweringConfigAttr::getWorkgroupReOrderingStrategy() const {
+  auto attrs = getAttributes();
+  if (auto attr = attrs.getAs<::mlir::iree_compiler::IREE::Codegen::
+                                  WorkgroupReorderingAttrInterface>(
+          kWorkgroupReOrderingStrategyName)) {
+    return attr;
+  }
+  return nullptr;
+}
+
+//===----------------------------------------------------------------------===//
+// Conditional Transpose Workgroup Reordering
+//===----------------------------------------------------------------------===//
+
+static std::tuple<SmallVector<OpFoldResult>, SmallVector<OpFoldResult>,
+                  SmallVector<OpFoldResult>>
+getLoopBounds(ArrayRef<Range> loopRanges,
+              ArrayRef<OpFoldResult> givenTileSizes) {
+  SmallVector<OpFoldResult> lbs, ubs, steps;
+  for (auto [loopRange, givenTileSize] :
+       llvm::zip_equal(loopRanges, givenTileSizes)) {
+    // No loop if the tile size is 0.
+    if (isZeroInteger(givenTileSize))
+      continue;
+    lbs.push_back(loopRange.offset);
+    ubs.push_back(loopRange.size);
+    steps.push_back(givenTileSize);
+  }
+  return {lbs, ubs, steps};
+}
+
+static Value computeNbTiles(OpBuilder &b, Location loc, OpFoldResult size,
+                            OpFoldResult tileSize) {
+  Value sizeVal = getValueOrCreateConstantIntOp(b, loc, size);
+  Value tileSizeVal = getValueOrCreateConstantIntOp(b, loc, tileSize);
+  auto nbTiles =
+      b.create<mlir::arith::DivUIOp>(loc, sizeVal, tileSizeVal).getResult();
+  return getValueOrCreateConstantIntOp(b, loc, nbTiles);
+}
+
+/// Computes the total number of different Tile loads accross all XCDs per
+/// iteration for the pingpong matmul kernel. For each iteration, we distribute
+/// tile loads to invidiual CUs along the X axis of the RHS.
+static Value computeNumTileLoads(OpBuilder &b, Location loc, OpFoldResult size,
+                                 OpFoldResult tileSize, Value nbXcds,
+                                 Value nbCus) {
+  Value nbTiles = computeNbTiles(b, loc, size, tileSize);
+  auto totalCUs = b.create<mlir::arith::MulIOp>(loc, nbXcds, nbCus);
+  auto nbX = b.create<mlir::arith::MinUIOp>(loc, nbTiles, totalCUs).getResult();
+  auto nbY = b.create<mlir::arith::CeilDivUIOp>(loc, totalCUs, nbX);
+  return getValueOrCreateConstantIndexOp(
+      b, loc, b.create<mlir::arith::AddIOp>(loc, nbX, nbY).getResult());
```

Comment:
I don't get the math here: if we have `numTiles >= totalCUs`, this always return `totalCUs` + 1?